sockeye/translate.py

```diff
@@ -55,52 +52,95 @@ def main():
-            args.softmax_temperature,
-            args.max_output_length_num_stds,
+        models, source_vocabs, target_vocab = inference.load_models(
+            context=context,
+            max_input_len=args.max_input_len,
+            beam_size=args.beam_size,
+            batch_size=args.batch_size,
+            model_folders=args.models,
+            checkpoints=args.checkpoints,
+            softmax_temperature=args.softmax_temperature,
+            max_output_length_num_stds=args.max_output_length_num_stds,
             decoder_return_logit_inputs=args.restrict_lexicon is not None,
             cache_output_layer_w_b=args.restrict_lexicon is not None)
-        restrict_lexicon = None # type: TopKLexicon
+        restrict_lexicon = None  # type: TopKLexicon
         if args.restrict_lexicon:
-            restrict_lexicon = TopKLexicon(vocab_source, vocab_target)
+            restrict_lexicon = TopKLexicon(source_vocabs[0], target_vocab)
             restrict_lexicon.load(args.restrict_lexicon)
-        translator = sockeye.inference.Translator(context,
-                                                  args.ensemble_mode,
-                                                  args.bucket_width,
-                                                  sockeye.inference.LengthPenalty(args.length_penalty_alpha,
-                                                                                  args.length_penalty_beta),
-                                                  models,
-                                                  vocab_source,
-                                                  vocab_target,
-                                                  restrict_lexicon)
-        read_and_translate(translator, output_handler, args.chunk_size, args.input)
-
-
-def read_and_translate(translator: sockeye.inference.Translator, output_handler: sockeye.output_handler.OutputHandler,
-                       chunk_size: Optional[int], source: Optional[str] = None) -> None:
+        translator = inference.Translator(context=context,
+                                          ensemble_mode=args.ensemble_mode,
+                                          bucket_source_width=args.bucket_width,
+                                          length_penalty=inference.LengthPenalty(args.length_penalty_alpha,
+                                                                                 args.length_penalty_beta),
+                                          models=models,
+                                          source_vocabs=source_vocabs,
+                                          target_vocab=target_vocab,
+                                          restrict_lexicon=restrict_lexicon)
+        read_and_translate(translator=translator,
+                           output_handler=output_handler,
+                           chunk_size=args.chunk_size,
+                           input=args.input,
+                           input_factors=args.input_factors,
```

Comment:
would it make sense to add a check for the number of input_factors?

Reply:
ok